File: src/server/src/main/java/com/project/domain/workout/Workout.java
package com.project.domain.workout;

import com.project.domain.BaseEntity;
import com.project.infrastructure.converters.DurationToIntervalConverter;
import com.project.utilities.FindUtil;
import com.project.utilities.Time;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToOne;

import java.time.Duration;
import java.time.ZonedDateTime;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Predicate;

@Entity(name = "workouts")
public class Workout extends BaseEntity {
    @Id
    private final UUID id;

    @OneToOne
    @JoinColumn(name = "workout_template_id")
    private WorkoutTemplate workoutTemplate;

    private ZonedDateTime finishedAt;

    @Convert(converter = DurationToIntervalConverter.class)
    private Duration duration;

    @Enumerated(EnumType.STRING)
    private WorkoutStatus status;

    protected Workout() {
        super();
        this.id = UUID.randomUUID();
        this.status = WorkoutStatus.Started;
    }

    public UUID getId() {
        return this.id;
    }

    public WorkoutTemplate getWorkoutTemplate() {
        return this.workoutTemplate;
    }

    public ZonedDateTime getFinishedAt() {
        return this.finishedAt;
    }

    public Duration getDuration() {
        return this.duration;
    }

    public WorkoutStatus getStatus() {
        return this.status;
    }

    public static Workout start(WorkoutTemplate workoutTemplate) {
        var workout = new Workout();
        workout.setWorkoutTemplate(workoutTemplate);
        return workout;
    }

    public static Workout startEmptyWorkout() {
        return new Workout();
    }

    public void finish() {
        this.status = WorkoutStatus.Finished;
        this.finishedAt = Time.utcNow();
        this.setDuration();
        this.setModifiedAt(Time.utcNow());
    }

    public void cancel() {
        this.status = WorkoutStatus.Cancelled;
        this.setModifiedAt(Time.utcNow());
    }

    public void addExercise(WorkoutExercise exercise) {
        if (this.workoutTemplate == null) {
            this.workoutTemplate = new WorkoutTemplate();
        }

        this.workoutTemplate.addExercise(exercise);
        this.setModifiedAt(Time.utcNow());
    }

    public void addSetToExercise(UUID exerciseId, double weight, int repetitions) {
        var exercise = this.findExerciseBy(exerciseId);
        exercise.ifPresent(e -> e.addSet(repetitions, weight));
        this.setModifiedAt(Time.utcNow());
    }

    public void updateExercise(UUID exerciseId, UUID setId, double weight, int repetitions) {
        var exercise = this.findExerciseBy(exerciseId);
        exercise.ifPresent(e -> e.updateSet(setId, weight, repetitions));

        this.setModifiedAt(Time.utcNow());
    }

    private Optional<WorkoutExercise> findExerciseBy(UUID exerciseId) {
        Predicate<WorkoutExercise> predicate = e -> e.getExerciseId().equals(exerciseId);
        return FindUtil.findByProperty(this.workoutTemplate.getExercises(), predicate);
    }

    private void setDuration() {
        this.duration = Duration.between(this.getCreatedAt(), this.finishedAt);
    }

    private void setWorkoutTemplate(WorkoutTemplate workoutTemplate) {
        this.workoutTemplate = workoutTemplate;
    }
}

File: src/server/src/main/kotlin/com/project/domain/workout/Workout.kt
package com.project.domain.workout

import com.project.common.BaseEntity
import com.project.infrastructure.converters.DurationToIntervalConverter
import com.project.utilities.FindUtil.findByProperty
import com.project.utilities.Time.utcNow
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToOne
import java.time.Duration
import java.time.Instant
import java.util.UUID

@Entity(name = "workouts")
class Workout protected constructor() : BaseEntity() {
    @Id
    val id: UUID = UUID.randomUUID()

    @OneToOne
    @JoinColumn(name = "workout_template_id")
    var workoutTemplate: WorkoutTemplate? = null

    var finishedAt: Instant? = null
        private set

    @Convert(converter = DurationToIntervalConverter::class)
    var duration: Duration? = null
        private set

    @Enumerated(EnumType.STRING)
    var status: WorkoutStatus
        private set

    init {
        this.status = WorkoutStatus.Started
    }

    fun finish() {
        this.status = WorkoutStatus.Finished
        this.finishedAt = utcNow()
        this.setDuration()
        this.modifiedAt = utcNow()
    }

    fun cancel() {
        this.status = WorkoutStatus.Cancelled
        this.modifiedAt = utcNow()
    }

    fun addExercise(exercise: WorkoutExercise) {
        if (this.workoutTemplate == null) {
            this.workoutTemplate = WorkoutTemplate()
        }

        workoutTemplate!!.addExercise(exercise)
        this.modifiedAt = utcNow()
    }

    fun addSetToExercise(exerciseId: UUID, weight: Double, repetitions: Int) {
        val exercise = this.findExerciseBy(exerciseId)
        exercise?.addSet(repetitions, weight)
        this.modifiedAt = utcNow()
    }

    fun updateExercise(exerciseId: UUID, setId: UUID, weight: Double, repetitions: Int) {
        val exercise = this.findExerciseBy(exerciseId)
        exercise?.updateSet(setId, weight, repetitions)

        this.modifiedAt = utcNow()
    }

    private fun findExerciseBy(exerciseId: UUID): WorkoutExercise? {
        val predicate = {e: WorkoutExercise? -> e?.exerciseId == exerciseId }
        return findByProperty(workoutTemplate?.exercises, predicate)
    }

    private fun setDuration() {
        this.duration = Duration.between(this.createdAt, this.finishedAt)
    }

    companion object {
        @JvmStatic
        fun start(workoutTemplate: WorkoutTemplate): Workout {
            val workout = Workout()
            workout.workoutTemplate = workoutTemplate
            return workout
        }

        @JvmStatic
        fun startEmptyWorkout(): Workout {
            return Workout()
        }
    }
}
